src/backend/src/main/java/com/sailsnap/backend/repositories/S3Repository.java

```diff
@@ -0,0 +1,125 @@
+                    .build();
+
+            return s3Client.listObjectsV2Paginator(request);
+        } catch (S3Exception e) {
+            log.error("Error listing objects in bucket '{}' with prefix '{}'", businessName, key, e);
+        }
+
+        return null;
+    }
+
+    /**
+     * Saves the file in S3 using the passed gallery name and business name.
+     * @param compressedStream, the image as an input stream
+     * @param galleryName, the name of the gallery that the business specified
+     * @param contentType, image or video
+     * @param contentLength, size of the compressedStream
+     * @param businessName, the name of the business, this is also their bucket name
+     * */
+    public void saveFile(InputStream compressedStream, String galleryName, String contentType, long contentLength, String businessName) {
+        String key = createKey(galleryName);
+        String sanitizedBucketName = sanitizeBucketName(businessName);
+
+        try {
+            log.info("Saving file to S3: {}", key);
+
+            PutObjectRequest request = PutObjectRequest.builder()
+                    .bucket(sanitizedBucketName)
+                    .key(key)
+                    .contentType(contentType)
+                    .contentLength(contentLength)
+                    .build();
+
+            s3Client.putObject(request, RequestBody.fromInputStream(compressedStream, contentLength));
+        } catch (S3Exception e) {
+            log.error("Error saving file to S3. Bucket: {}, Key: {}", businessName, key, e);
+        }
+    }
+
+    /**
+     * This method uses the business, gallery name. As well as the date and time to create the key for batch-saving
+     * media files. For example, if lakers made a gallery called okanagan-lake, on July 5th 2025 the key would be:
+     * lakers/2025/07/05/okanagan-lake. Then all the files would be stored there
+     * */
+    private String createKey(String galleryName) {
+        LocalDateTime now = LocalDateTime.now();
+
+        DateTimeFormatter formatter = DateTimeFormatter.ofPattern("yyyy/MM/dd");
+        String formattedDateTime = now.format(formatter);
+
+        return String.format("%s/%s", formattedDateTime, galleryName);
+    }
+
+    /**
+     * Sanitize bucket name to adhere to s3 conventions
+     * */
+    private String sanitizeBucketName(String businessName) {
+        String sanitizedBucketName = businessName.toLowerCase()
+                .replaceAll("[^a-z0-9.-]", "-")
+                .replaceAll("-{2,}", "-");
+
+        sanitizedBucketName = sanitizedBucketName.replaceAll("^[.-]+", "").replaceAll("[.-]+$", "");
+        if (sanitizedBucketName.length() < 3) {
+            sanitizedBucketName = sanitizedBucketName + "-".repeat(3 - sanitizedBucketName.length());
+        } else if (sanitizedBucketName.length() > 63) {
+            sanitizedBucketName = sanitizedBucketName.substring(0, 63).replaceAll("[.-]+$", "");
+        }
+        return sanitizedBucketName;
```

Comment:
Should a random UUID be added to guarantee uniqueness? Such as:  
```suggestion
       sanitizedBucketName += "-" + UUID.randomUUID().toString().substring(0, 8);
```

Reply:
you're right, it might be even better to use the a random uuid entirely so it's unique, and just store the uuid in the table with their other data